### PR TITLE
More g UI and some configuration changes

### PR DESCRIPTION
**Pull request by MickyHoltzman**

### CODE REVIEW

File: modelling_opponent/OpponentType.py
Comment (Anna4142) [inline modelling_opponent/OpponentType.py:1]: we will need more classes for fixed strategy and learners thats of type enum?

File: Experiment_Launcher_code/ExperimentLauncher.py [hunk 1]
```diff
@@ -6,27 +15,34 @@ def main():
...
-        num_trials = settings.get('num_trials')
-        return_time = settings.get('return_time')
-        decision_time = settings.get('decision_time')
-        opponent_type = settings.get('opponent_type')
-        opponent1_strategy = settings.get('opponent1_strategy')
-        opponent2_strategy = settings.get('opponent2_strategy')
-        opponent1_probability = settings.get('opponent1_probability')
-        opponent2_probability = settings.get('opponent2_probability')
+    # After the GUI is closed, get the settings using the appropriate methods
+    if experiment_gui.is_data_valid():
```
Comment (Anna4142) [inline Experiment_Launcher_code/ExperimentLauncher.py:19]: why not check validity before passing to thr experimentlauncher in the gui program itself?

File: Experiment_Launcher_code/ExperimentLauncher.py [hunk 2]
```diff
@@ -6,27 +15,34 @@ def main():
...
+        comport_name = experiment_gui.get_com_port()
+        experiment_parameters = experiment_gui.get_experiment_parameters()
+        opponent_configuration = experiment_gui.get_opponent_configuration()
+
+        # Instantiate software components
+        video_analyzer = Video_Analyzer("1780")
+        reward_manager = RewardManager(comport_name)
+
+        if opponent_configuration.get("opponent1_type") == OpponentType.MOUSE:
```
Comment (Anna4142) [inline Experiment_Launcher_code/ExperimentLauncher.py:28]: i assume youve moved this part from the experimentlauncher to the manager as you plan to incorporate further changes  regarding the types of learners and strategies

File: Experiment_Launcher_code/ExperimentLauncher.py [hunk 3]
```diff
@@ -6,27 +15,34 @@ def main():
...
-        opponent1_probability = settings.get('opponent1_probability')
-        opponent2_probability = settings.get('opponent2_probability')
+    # After the GUI is closed, get the settings using the appropriate methods
+    if experiment_gui.is_data_valid():
+        comport_name = experiment_gui.get_com_port()
+        experiment_parameters = experiment_gui.get_experiment_parameters()
+        opponent_configuration = experiment_gui.get_opponent_configuration()
+
+        # Instantiate software components
+        video_analyzer = Video_Analyzer("1780")
```
Comment (Anna4142) [inline Experiment_Launcher_code/ExperimentLauncher.py:26]: i dont understand why you want to initialize the video analyser  and reward manager here.seemed better in experiment_manager

File: Experiment_Launcher_code/ExperimentLauncher.py [hunk 4]
```diff
@@ -6,27 +15,34 @@ def main():
...
+            first_opponent = FixedStrategyPrisoner(opponent_configuration.get("opponent1_strategy"), opponent_configuration.get("opponent1_probability"))
+        else:
+            first_opponent = Simulated_mouse()
+
+        if opponent_configuration.get("opponent2_type") == OpponentType.MOUSE:
+            second_opponent = MouseMonitor(2, video_analyzer, reward_manager)
+        elif opponent_configuration.get("opponent2_type") == OpponentType.FIXED_STRATEGY:
+            second_opponent = FixedStrategyPrisoner(opponent_configuration.get("opponent2_strategy"), opponent_configuration.get("opponent2_probability"))
+        else:
+            second_opponent = Simulated_mouse()
```
Comment (Anna4142) [inline Experiment_Launcher_code/ExperimentLauncher.py:41]: why did you move this section from the experiment manager to here?

File: Experiment_Launcher_code/ExperimentLauncher.py [hunk 5]
```diff
@@ -6,27 +15,34 @@ def main():
...
 
         # Initialize and start the experiment
-        expManager = ExperimentManager(comport_name)
+        expManager = ExperimentManager(video_analyzer, reward_manager)
```
Comment (Anna4142) [inline Experiment_Launcher_code/ExperimentLauncher.py:44]: ok but why

File: Experiment_Launcher_code/experimentgui.py [hunk 6]
```diff
@@ -177,25 +157,35 @@ def get_settings(self):
...
-
-            #print(settings)
-            return settings
 
         except ValueError:
             # Handle the case where the input is not a valid integer
             messagebox.showerror("Invalid Input", "Please enter valid numbers for trials, duration, and decision time.")
             return None
 
+    def is_data_valid(self):
```
Comment (Anna4142) [inline Experiment_Launcher_code/experimentgui.py]: seems more logical to check the trial validity in the experiment gui itself instead of in the experiment launcher to me.please can you explain this change

**Review by Anna4142 — COMMENTED** [review]
please have  look at y comments

File: Experiment_Launcher_code/ExperimentLauncher.py (hunk 3 above)
Comment (MickyHoltzman) [reply Experiment_Launcher_code/ExperimentLauncher.py:26]: It's a question of system design and It could be done both ways, I agree.
Here is my line of thought:
a) The way it was previously written the ExperimentLauncher did not really do anything. It just took all the parameters from GUI and passed them to the experiment manager. So why do we need the Launcher class? 
b) On the other hand the ExperimentManager does not really care which opponents are  configured for for the experiment. Since all opponents are derived from the PrisonerABC class, the ExperimentManager behaves exactly the same way for all configurations. So why bother it with the knowledge of opponents types and with the task of configuring the opponents? Remember my moto: In a good system design knowledge is distributed on a need to know basis :-)
c) Therefore I decided to make the Launcher class responsible for opponent configuration. As a result it is the only class aware of opponents types. I think this is a good, solid, design decision we should stick too.
d) But, it created a kind of a side effect. VideoAnalyzer and RewardManager are used by the experiment_manager and mouse_monitor but not the other types of opponents. If we keep the knowledge of the differences in the opponent types (and their __init__() functions) within the ExperimentLauncher it needs to be the one who initializes the VideoAnalyzer and the RewardManager. The other option is to add a setVideoAnalyzer() and setRewardManager() methods to the prisonerABC class - and have the experimentManger initialize and set() these components.  I decided to bestow the responsibility of creating all SW components on the ExperimentLauncher. The other way will work too.

File: Experiment_Launcher_code/ExperimentLauncher.py (hunk 4 above)
Comment (MickyHoltzman) [reply Experiment_Launcher_code/ExperimentLauncher.py:41]: I explained this in my response to your previous comment

File: Experiment_Launcher_code/ExperimentLauncher.py (hunk 5 above)
Comment (MickyHoltzman) [reply Experiment_Launcher_code/ExperimentLauncher.py:44]: I explained this in my response to your previous comment

File: Experiment_Launcher_code/experimentgui.py (hunk 6 above)
Comment (MickyHoltzman) [reply Experiment_Launcher_code/experimentgui.py]: Agreed. But .... :-)
Data Validity is really checked in the GUI. As you initially wrote it, the StartExperiment button will not close the GUI window and return to the Launcher unless the data is valid. The problem is, that there is another way to close the GUI window, which is to click on the CloseWindow (x in the corner) button. This bypasses the validity check. I did not want to disable this option since the experimenter needs to have a way to change their mind and stop the experiment, before it even started. 
Bottom line, I needed a way to differentiate between StartExperiment and CloseWindow and I used the data validity. You are right, this is a bit awkward. let me Fixed it.

File: modelling_opponent/OpponentType.py
Comment (MickyHoltzman) [reply modelling_opponent/OpponentType.py:1]: We might. Right now the enum covers only the difference between mouse, FixedStrategy and Learner. The various strategies and learners are using the radio button strings rather being concerted to enum. We should be consistent, I agree. 
I can see three options:
(a) get rid of the enum and use the radio button strings for the opponent type (the first radio button on GUI)
(b) add all opponent types to this enum.
(c) create dedicated enums for Strategy and Learner.
Which do you prefer?

File: Experiment_Launcher_code/experimentgui.py (hunk 6 above)
Comment (MickyHoltzman) [reply Experiment_Launcher_code/experimentgui.py]: Fixed. Please look at the new code version.

File: Experiment_Launcher_code/ExperimentLauncher.py (hunk 3 above)
Comment (Anna4142) [reply Experiment_Launcher_code/ExperimentLauncher.py:26]: I agree with a) b) and c) .It makes sense to create the agents in the launcher but I feel like we might be making the PrisonerABC class unnecesarily complex if we add the setVideoAnalyzer() and setRewardManager() methods as they are only use by the mouse monitor and not the fixed strategy class or the learners

File: modelling_opponent/OpponentType.py
Comment (Anna4142) [reply modelling_opponent/OpponentType.py:1]: i like using enums as its cleaner than using radio button strings. but i dont want to clutter a single enum class with all the options so i feel option c is best

**Review by Anna4142 — COMMENTED** [review]
ive added comments.please have a look

File: Experiment_Launcher_code/ExperimentLauncher.py (hunk 3 above)
Comment (MickyHoltzman) [reply Experiment_Launcher_code/ExperimentLauncher.py:26]: So, basically you agree with (d) as well :-). 

If we are creating the agents in the launcher and we want the video analyzer and reward manager to be set for the MouseMonitor through its __init__() function  (which is unique to the MouseMonitor and does not require a change in PrisonerABC) we need to instantiate the two SW components in Launcher as well - which is what I did. Also since the ExperimentManager is also using both of them it is set for it as well.

File: modelling_opponent/OpponentType.py
Comment (MickyHoltzman) [reply modelling_opponent/OpponentType.py:1]: You got it. Can it please wait for next PR. I have other GUI changes I need to do as well (directory structure and, mouse ID, etc...)

Comment (Anna4142) [reply modelling_opponent/OpponentType.py:1]: yes sure.i think itll be best to do all the fetching to my local when you come to the lab

File: Experiment_Launcher_code/ExperimentLauncher.py (hunk 3 above)
Comment (Anna4142) [reply Experiment_Launcher_code/ExperimentLauncher.py:26]: yes i think this change sounds good

**Review by Anna4142 — APPROVED** [review]

**Comment by MickyHoltzman** [comment]
I made too many changes. So I am closing the PR and opening a new one.
